### PR TITLE
Updated search requests

### PR DESCRIPTION
**Pull request by mhilton**
Reworked the elasticserch DSL generation to be simpler. Make use of ngrams for autocomplete search.

### CODE REVIEW

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results: http://ci.jujugui.org:8080/job/charmstore/370/

**Comment by jrwren** [comment]
:+1:

File: internal/charmstore/essettings_test.go [hunk 1]
```diff
@@ -0,0 +1,284 @@
...
+                "type": "string"
+              },
+              "Default": {
+                "type": "string"
+              }
+            }
+          }
+        }
+      },
+      "charmactions": {
```
Comment (bac) [inline internal/charmstore/essettings_test.go]: Camel case CharmActions and components

File: internal/elasticsearch/elasticsearch.go [hunk 2]
```diff
@@ -102,6 +102,46 @@ func (db *Database) PutDocument(index, type_, id string, doc interface{}) error
 	return nil
 }
 
+// PutIndex creates or updates the index with the given configuration
+func (db *Database) PutIndex(index string, config interface{}) error {
```
Comment (bac) [inline internal/elasticsearch/elasticsearch.go:106]: Please punctate all comments.

File: internal/elasticsearch/query.go [hunk 3]
```diff
@@ -3,54 +3,133 @@
...
+}
+
+// BoostField creates a string which represents a field name with a boost value.
+func BoostField(field string, boost float64) string {
+	return fmt.Sprintf("%s^%f", field, boost)
 }
 
 // MatchAllQuery provides a query that matches all
-// documents in the index.
+// documents in the index
```
Comment (bac) [inline internal/elasticsearch/query.go]: :neckbeard:

File: internal/elasticsearch/query.go [hunk 4]
```diff
@@ -3,54 +3,133 @@
...
+
+// FunctionScoreQuery provides a query that includes
+func (f FunctionScoreQuery) MarshalJSON() ([]byte, error) {
+	return marshalJSON("function_score", map[string]interface{}{
+		"query":     f.Query,
+		"functions": f.Functions,
+	})
+}
+
+// Function is a function definition for use with a FunctionScoreQuery
```
Comment (bac) [inline internal/elasticsearch/query.go]: :anguished:

File: internal/elasticsearch/query.go [hunk 5]
```diff
@@ -3,54 +3,133 @@
...
+	return marshalJSON("match_all", struct{}{})
+}
+
 // MatchQuery provides a query that matches against
 // a complete field.
-type MatchQuery map[string]interface{}
+type MatchQuery struct {
+	Field string
+	Query string
+}
```
Comment (bac) [inline internal/elasticsearch/query.go:42]: Thanks for this.

File: internal/elasticsearch/query.go [hunk 6]
```diff
@@ -3,54 +3,133 @@
...
 
-// PrefixQuery provides a query that matches against
-// the start of a field.
-type PrefixQuery map[string]interface{}
+func (m MatchQuery) MarshalJSON() ([]byte, error) {
+	return marshalJSON("match", map[string]interface{}{m.Field: m.Query})
```
Comment (bac) [inline internal/elasticsearch/query.go:45]: return marshalJSON("match", marshalJSON(m.Field, m.Query))

???

**Comment by bac** [comment]
:+1:

File: internal/charmstore/essettings_test.go (hunk 1 above)
Comment (mhilton) [inline internal/charmstore/essettings_test.go]: done.

File: internal/elasticsearch/elasticsearch.go (hunk 2 above)
Comment (mhilton) [inline internal/elasticsearch/elasticsearch.go:106]: done

File: internal/elasticsearch/query.go (hunk 3 above)
Comment (mhilton) [inline internal/elasticsearch/query.go]: done

File: internal/elasticsearch/query.go (hunk 4 above)
Comment (mhilton) [inline internal/elasticsearch/query.go]: done

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results: http://ci.jujugui.org:8080/job/charmstore/371/

**Comment by mhilton** [comment]
:shipit:

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/charmstore-merge

**Pull request merged**